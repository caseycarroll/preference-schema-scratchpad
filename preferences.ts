import { toFashionPreferences, toPreferencesWithDescriptors } from "./utils.ts";

// Shape of user preferences in CDP
export type ExplicitPreference = {
    value: string;
    categoryId: number;
}

// Introduce a type for grouping preferences into generic categories
interface FocusCateogryGroup {
    categoryId: number; // Lower level and generic category id i.e. Women's Pants
    name: string; // Name of the category i.e. Women's Pants
    categoryIds: number[]; // List of category ids that belong to this category
}

// Introduce a new type for focus category preferences
interface FocusCategoryPreference {
    value: string;
    focusCategoryGroup: FocusCateogryGroup;
}

// Extend FocusCateogryGrouping for Fashion use case
interface FashionFocusCategoryGroup extends FocusCateogryGroup {
    // add gender to avoid string parsing of name in base
    // useful for grouping data in the UI
    gender: 'all' | 'men' | 'women';
    type: string; // Type of clothing i.e. Pants, Shirt
}

// Extend FocusCategoryPreference for Fashion use case
export interface FashionPreference extends FocusCategoryPreference {
    descriptor?: string;
    focusCategoryGroup: FashionFocusCategoryGroup;
};

// User preferences in CDP
const userPreferences: ExplicitPreference[] = [
    { value: 'large,tall|medium', categoryId: 1 },
    { value: 'medium', categoryId: 2 },
    { value: 'small', categoryId: 3 },
    { value: '32,waist', categoryId: 4 },
    { value: '34,waist', categoryId: 5 },
    { value: '36,inseam', categoryId: 6 },
    { value: '38', categoryId: 7 },
    { value: 'large', categoryId: 10 },
    { value: 'medium', categoryId: 9 },
    { value: 'adidas', categoryId: 11 },
];

// Fashion Category Groupings maintained by humans
// Imagine this comes from a config
export const fashionCategories: FashionFocusCategoryGroup[] = [
    { categoryIds: [1, 2, 3], name: `Men's Pants`, gender: 'men', type: 'pants', categoryId: 10 },
    { categoryIds: [4, 5, 6, 7, 8], name: `Women's Pants`, gender: 'women', type: 'pants', categoryId: 20 },
    { categoryIds: [9, 10], name: `Men's Shirts`, gender: 'men', type: 'shirt', categoryId: 30 },
    { categoryIds: [11], name: `Brands`, gender: 'all', type: 'brands', categoryId: 40 },
];

function resolveFashionPreferences(): FashionPreference[] {
    return userPreferences.reduce(toFashionPreferences, []).reduce(toPreferencesWithDescriptors, []);
}

// Imaginary GraphQL resolvers on Stashplex
export const resolvers = {
    Query: {
        preferences: () => userPreferences,
        fashionCategories: () => fashionCategories,
        fashionPreferences: () => resolveFashionPreferences(),
    },
}