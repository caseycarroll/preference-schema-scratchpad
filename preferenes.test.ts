import { describe, expect, it } from 'vitest';
import { resolvers } from './preferences';

describe('Preference Resolvers', () => {
    it('should resolve fashion preferences', () => {
        const fashionPreferences = resolvers.Query.fashionPreferences();
        expect(fashionPreferences).toMatchSnapshot();
    });
    it('should resolve fashion preferences by type', () => {
        const fashionPreferencesByType = resolvers.Query.fashionPreferencesByType();
        expect(fashionPreferencesByType).toMatchSnapshot();
    });
    it('should resolve fashion preferences by type and gender', () => {
        const fashionPreferencesByTypeAndGender = resolvers.Query.fashionPrefrencesByTypeAndGender();
        expect(fashionPreferencesByTypeAndGender).toMatchSnapshot();  
    });
});
