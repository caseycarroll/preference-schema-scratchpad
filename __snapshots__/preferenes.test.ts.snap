// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Preference Resolvers > should resolve fashion preferences 1`] = `
[
  {
    "descriptor": "tall",
    "focusCategoryGroup": {
      "categoryId": 10,
      "categoryIds": [
        1,
        2,
        3,
      ],
      "gender": "men",
      "name": "Men's Pants",
      "type": "pants",
    },
    "leafCat": 1,
    "value": "large",
  },
  {
    "descriptor": undefined,
    "focusCategoryGroup": {
      "categoryId": 10,
      "categoryIds": [
        1,
        2,
        3,
      ],
      "gender": "men",
      "name": "Men's Pants",
      "type": "pants",
    },
    "leafCat": 1,
    "value": "medium",
  },
  {
    "descriptor": undefined,
    "focusCategoryGroup": {
      "categoryId": 10,
      "categoryIds": [
        1,
        2,
        3,
      ],
      "gender": "men",
      "name": "Men's Pants",
      "type": "pants",
    },
    "leafCat": 2,
    "value": "medium",
  },
  {
    "descriptor": undefined,
    "focusCategoryGroup": {
      "categoryId": 10,
      "categoryIds": [
        1,
        2,
        3,
      ],
      "gender": "men",
      "name": "Men's Pants",
      "type": "pants",
    },
    "leafCat": 3,
    "value": "small",
  },
  {
    "descriptor": "waist",
    "focusCategoryGroup": {
      "categoryId": 20,
      "categoryIds": [
        4,
        5,
        6,
        7,
        8,
      ],
      "gender": "women",
      "name": "Women's Pants",
      "type": "pants",
    },
    "leafCat": 4,
    "value": "32",
  },
  {
    "descriptor": "waist",
    "focusCategoryGroup": {
      "categoryId": 20,
      "categoryIds": [
        4,
        5,
        6,
        7,
        8,
      ],
      "gender": "women",
      "name": "Women's Pants",
      "type": "pants",
    },
    "leafCat": 5,
    "value": "34",
  },
  {
    "descriptor": "inseam",
    "focusCategoryGroup": {
      "categoryId": 20,
      "categoryIds": [
        4,
        5,
        6,
        7,
        8,
      ],
      "gender": "women",
      "name": "Women's Pants",
      "type": "pants",
    },
    "leafCat": 6,
    "value": "36",
  },
  {
    "descriptor": undefined,
    "focusCategoryGroup": {
      "categoryId": 20,
      "categoryIds": [
        4,
        5,
        6,
        7,
        8,
      ],
      "gender": "women",
      "name": "Women's Pants",
      "type": "pants",
    },
    "leafCat": 7,
    "value": "38",
  },
  {
    "descriptor": undefined,
    "focusCategoryGroup": {
      "categoryId": 30,
      "categoryIds": [
        9,
        10,
      ],
      "gender": "men",
      "name": "Men's Shirts",
      "type": "shirt",
    },
    "leafCat": 10,
    "value": "large",
  },
  {
    "descriptor": undefined,
    "focusCategoryGroup": {
      "categoryId": 30,
      "categoryIds": [
        9,
        10,
      ],
      "gender": "men",
      "name": "Men's Shirts",
      "type": "shirt",
    },
    "leafCat": 9,
    "value": "medium",
  },
  {
    "descriptor": undefined,
    "focusCategoryGroup": {
      "categoryId": 40,
      "categoryIds": [
        11,
      ],
      "gender": "all",
      "name": "Brands",
      "type": "brands",
    },
    "leafCat": 11,
    "value": "adidas",
  },
]
`;
